tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc

```diff
@@ -82,44 +103,114 @@ static void MaybeAppendSlash(std::string* name) {
     name->push_back('/');
 }
 
+// A helper function to actually read the data from GCS.
+static int64_t LoadBufferFromGCS(const std::string& path, size_t offset,
+                                 size_t buffer_size, char* buffer,
+                                 gcs::Client* gcs_client, TF_Status* status) {
+  std::string bucket, object;
+  ParseGCSPath(path, false, &bucket, &object, status);
+  if (TF_GetCode(status) != TF_OK) return -1;
+  auto stream = gcs_client->ReadObject(
+      bucket, object, gcs::ReadRange(offset, offset + buffer_size));
+  TF_SetStatusFromGCSStatus(stream.status(), status);
+  if ((TF_GetCode(status) != TF_OK) &&
+      (TF_GetCode(status) != TF_OUT_OF_RANGE)) {
+    return -1;
+  }
+  int64_t read;
+  if (!absl::SimpleAtoi(stream.headers().find("content-length")->second,
+                        &read)) {
+    TF_SetStatus(status, TF_UNKNOWN, "Could not get content-length header");
+    return -1;
+  }
+  stream.read(buffer, read);
+  return stream.gcount();
+}
+
 // SECTION 1. Implementation for `TF_RandomAccessFile`
 // ----------------------------------------------------------------------------
 namespace tf_random_access_file {
+using ReadFn =
+    std::function<int64_t(const std::string& path, uint64_t offset, size_t n,
+                          char* buffer, TF_Status* status)>;
 typedef struct GCSFile {
-  const std::string bucket;
-  const std::string object;
-  gcs::Client* gcs_client;  // not owned
+  const std::string path;
+  const bool is_cache_enable;
+  const uint64_t buffer_size;
+  ReadFn read_fn;
+  absl::Mutex buffer_mutex;
+  uint64_t buffer_start ABSL_GUARDED_BY(buffer_mutex);
+  bool buffer_end_is_past_eof ABSL_GUARDED_BY(buffer_mutex);
+  std::string buffer ABSL_GUARDED_BY(buffer_mutex);
+  GCSFile(std::string path, bool is_cache_enable, uint64_t buffer_size,
+          ReadFn read_fn)
+      : path(path),
+        is_cache_enable(is_cache_enable),
+        buffer_size(buffer_size),
+        buffer_mutex(),
+        buffer_start(0),
+        buffer_end_is_past_eof(false),
+        buffer(),
+        read_fn(std::move(read_fn)) {}
 } GCSFile;
 
 void Cleanup(TF_RandomAccessFile* file) {
   auto gcs_file = static_cast<GCSFile*>(file->plugin_file);
   delete gcs_file;
 }
 
-// TODO(vnvo2409): Adding cache.
+static void FillBuffer(uint64_t start, GCSFile* gcs_file, TF_Status* status) {
+  gcs_file->buffer_start = start;
+  gcs_file->buffer.resize(gcs_file->buffer_size);
+  auto read =
+      gcs_file->read_fn(gcs_file->path, gcs_file->buffer_start,
+                        gcs_file->buffer_size, &(gcs_file->buffer[0]), status);
+  gcs_file->buffer_end_is_past_eof = (TF_GetCode(status) == TF_OUT_OF_RANGE);
+  if (read >= 0) gcs_file->buffer.resize(read);
```

Comment:
```
tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:164:13: error: writing variable 'buffer_start' requires holding mutex 'gcs_file->buffer_mutex' exclusively [-Werror,-Wthread-safety-analysis]
  gcs_file->buffer_start = start;
            ^
tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:165:13: error: reading variable 'buffer' requires holding mutex 'gcs_file->buffer_mutex' [-Werror,-Wthread-safety-analysis]
  gcs_file->buffer.resize(gcs_file->buffer_size);
            ^
tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:167:51: error: reading variable 'buffer_start' requires holding mutex 'gcs_file->buffer_mutex' [-Werror,-Wthread-safety-analysis]
      gcs_file->read_fn(gcs_file->path, gcs_file->buffer_start,
                                                  ^
tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:168:60: error: reading variable 'buffer' requires holding mutex 'gcs_file->buffer_mutex' [-Werror,-Wthread-safety-analysis]
                        gcs_file->buffer_size, &(gcs_file->buffer[0]), status);
                                                           ^
tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:169:13: error: writing variable 'buffer_end_is_past_eof' requires holding mutex 'gcs_file->buffer_mutex' exclusively [-Werror,-Wthread-safety-analysis]
  gcs_file->buffer_end_is_past_eof = (TF_GetCode(status) == TF_OUT_OF_RANGE);
            ^
tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:170:28: error: reading variable 'buffer' requires holding mutex 'gcs_file->buffer_mutex' [-Werror,-Wthread-safety-analysis]
  if (read >= 0) gcs_file->buffer.resize(read);
                           ^
tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:468:25: error: reading variable 'file_block_cache' requires holding mutex 'gcs_file->block_cache_lock' [-Werror,-Wthread-safety-analysis]
            ? gcs_file->file_block_cache->Read(path, offset, n, buffer, status)
```

Reply:
The errors with `gcs_file->buffer_mutex` are caused by
https://github.com/tensorflow/tensorflow/blob/c4a07915a2868303edf496b32ae675bcbc7a0715/tensorflow/core/platform/cloud/gcs_file_system.cc#L352-L353
I forgot to add `ABSL_EXCLUSIVE_LOCKS_REQUIRED`.

The error with `gcs_file->block_cache_lock` is caused by https://github.com/tensorflow/tensorflow/blob/c4a07915a2868303edf496b32ae675bcbc7a0715/tensorflow/core/platform/cloud/gcs_file_system.cc#L992

Both are fixed now.